packages/repo/src/car.ts

```diff
@@ -0,0 +1,200 @@
+import * as cbor from '@ipld/dag-cbor'
+import { CID } from 'multiformats/cid'
+import * as ui8 from 'uint8arrays'
+import * as varint from 'varint'
+import {
+  check,
+  schema,
+  streamToBuffer,
+  verifyCidForBytes,
+} from '@atproto/common'
+import { parseCidFromBytes } from '@atproto/common/src'
+import { BlockMap } from './block-map'
+import { CarBlock } from './types'
+
+export async function* writeCarStream(
+  root: CID | null,
+  blocks: AsyncIterable<CarBlock>,
+): AsyncIterable<Uint8Array> {
+  const header = new Uint8Array(
+    cbor.encode({
+      version: 1,
+      roots: root ? [root] : [],
+    }),
+  )
+  yield new Uint8Array(varint.encode(header.byteLength))
+  yield header
+  for await (const block of blocks) {
+    yield new Uint8Array(
+      varint.encode(block.cid.bytes.byteLength + block.bytes.byteLength),
+    )
+    yield block.cid.bytes
+    yield block.bytes
+  }
+}
+
+export const blocksToCarFile = (
+  root: CID | null,
+  blocks: BlockMap,
+): Promise<Uint8Array> => {
+  const carStream = blocksToCarStream(root, blocks)
+  return streamToBuffer(carStream)
+}
+
+export const blocksToCarStream = (
+  root: CID | null,
+  blocks: BlockMap,
+): AsyncIterable<Uint8Array> => {
+  return writeCarStream(root, iterateBlocks(blocks))
+}
+
+async function* iterateBlocks(blocks: BlockMap) {
+  for (const entry of blocks.entries()) {
+    yield { cid: entry.cid, bytes: entry.bytes }
+  }
+}
+
+export const readCar = async (
+  bytes: Uint8Array,
+): Promise<{ roots: CID[]; blocks: BlockMap }> => {
+  const { roots, blocks } = await readCarStream([bytes])
+  const blockMap = new BlockMap()
+  for await (const block of blocks) {
+    blockMap.set(block.cid, block.bytes)
+  }
+  return { roots, blocks: blockMap }
+}
+
+export const readCarWithRoot = async (
+  bytes: Uint8Array,
+): Promise<{ root: CID; blocks: BlockMap }> => {
+  const { roots, blocks } = await readCar(bytes)
+  if (roots.length !== 1) {
+    throw new Error(`Expected one root, got ${roots.length}`)
+  }
+  const root = roots[0]
+  return {
+    root,
+    blocks,
+  }
+}
+
+export const readCarStream = async (
+  car: Iterable<Uint8Array> | AsyncIterable<Uint8Array>,
+): Promise<{
+  roots: CID[]
+  blocks: AsyncIterable<CarBlock>
+  close: () => Promise<void>
+}> => {
+  const reader = new BufferedReader(car)
+  try {
+    const headerSize = await reader.readVarint()
+    if (headerSize === null) {
+      throw new Error('Could not parse CAR header')
+    }
+    const headerBytes = await reader.read(headerSize)
+    const header = cbor.decode(headerBytes)
+    if (!check.is(header, schema.carHeader)) {
+      throw new Error('Could not parse CAR header')
+    }
+    return {
+      roots: header.roots,
+      blocks: readCarBlocksIter(reader),
+      close: reader.close,
```

Comment:
You need to bind the context ;-)

```suggestion
      close: () => reader.close(),
```